Make SpeechSynthesisUtterance an ActiveDOMObject

szewai opened the pull request:
#### 42ffbe195e69ce41e7c37707231efc7ed5e9b460
<pre>
Make SpeechSynthesisUtterance an ActiveDOMObject
<a href="https://bugs.webkit.org/show_bug.cgi?id=260214">https://bugs.webkit.org/show_bug.cgi?id=260214</a>
rdar://113921763

Reviewed by Ryosuke Niwa.

Currently dispatching event to SpeechSynthesisUtterance can lead to crash due to GC. To fix that, we now keep wrapper
of SpeechSynthesisUtterance alive until there is no more event to be dispatched. This can happen in two cases:
1. SpeechSynthesisUtterance is not processed yet, and is removed from SpeechSynthesis&apos;s queue (i.e. the utterance will
not be processed).
2. SpeechSynthesisUtterance is being processed, and it finishes dispatching end or error event -- these two events are
the last events to be fired on an untterance according to spec: <a href="https://wicg.github.io/speech-api/#utterance-events.">https://wicg.github.io/speech-api/#utterance-events.</a>

* Source/WebCore/Modules/speech/SpeechSynthesis.cpp:
(WebCore::SpeechSynthesis::setPlatformSynthesizer):
(WebCore::SpeechSynthesis::cancel):
(WebCore::SpeechSynthesis::clearUtteranceQueue):
* Source/WebCore/Modules/speech/SpeechSynthesis.h:
* Source/WebCore/Modules/speech/SpeechSynthesisUtterance.cpp:
(WebCore::SpeechSynthesisUtterance::SpeechSynthesisUtterance):
(WebCore::SpeechSynthesisUtterance::eventOccurred):
(WebCore::SpeechSynthesisUtterance::errorEventOccurred):
(WebCore::SpeechSynthesisUtterance::dispatchEventAndUpdateState):
(WebCore::SpeechSynthesisUtterance::setIsActiveForEventDispatch):
(WebCore::SpeechSynthesisUtterance::activeDOMObjectName const):
(WebCore::SpeechSynthesisUtterance::virtualHasPendingActivity const):
* Source/WebCore/Modules/speech/SpeechSynthesisUtterance.h:
* Source/WebCore/Modules/speech/SpeechSynthesisUtterance.idl:

Canonical link: <a href="https://commits.webkit.org/266959@main">https://commits.webkit.org/266959@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/48f207c85a73f471a470e3a293d0ebafdea60a06

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/15232 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/15537 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/15897 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/16988 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/14309 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/15372 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/18052 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/15635 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/16920 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/15415 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/15854 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/12953 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/17721 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/13134 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/13741 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/20703 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/14214 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/13909 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/17166 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/14477 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/12270 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/13752 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3660 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/18096 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/14314 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/Modules/speech/SpeechSynthesisUtterance.cpp:
```diff
@@ -81,26 +81,49 @@ void SpeechSynthesisUtterance::setVoice(SpeechSynthesisVoice* voice)
+    setIsActive(false);
+}
+
+void SpeechSynthesisUtterance::setIsActive(bool isActive)
```
Can we use activity token instead of this boolean? (ActiveDOMObject::makePendingActivity)

szewai replied:
Do you mean to make a `PendingActivity` on `setIsActive(true)` and clear it when `setIsActive(false)`? 
Then how do we check event listeners (and their changes)? Or do you mean we should keep object alive even when there's no event listener?

rniwa replied:
I don't think we need to check whether there is an event listener or not. Keeping it alive whenever there is a pending event is okay.

szewai replied:
Ok, updated patch based on comment

rniwa commented on Source/WebCore/Modules/speech/SpeechSynthesis.cpp:
```diff
@@ -151,6 +152,7 @@ void SpeechSynthesis::speak(SpeechSynthesisUtterance& utterance)
         return;
 #endif
 
+    utterance.setIsActiveForEventDispatch(true);
```
oh, I meant that SpeechSynthesis would call makeActivity on utterance and store it there.
That way, you can clear that vector in addition to m_utteranceQueue and we're good to go.
But now I see that SpeechSynthesisUtterance has to clear this activity as well.
In that case, I think we can go back to your original approach. It's simpler & easier to follow.

szewai replied:
Oh. `SpeechSynthesis` does not uniquely own `SpeechSynthesisUtterance` (see `VTTCue::prepareToSpeak` for example), so clearing the vector won't make `SpeechSynthesisUtterance` go away.
Ok I will go with the previous approach as in https://github.com/WebKit/WebKit/commit/472814b941249f41ca2c1c89888008dfaa8ce3e6. Thanks for the review!

rniwa reviewed (approved):
r=me on the original PR.

szewai committed:
Make SpeechSynthesisUtterance an ActiveDOMObject
https://bugs.webkit.org/show_bug.cgi?id=260214
rdar://113921763

Reviewed by Ryosuke Niwa.

Currently dispatching event to SpeechSynthesisUtterance can lead to crash due to GC. To fix that, we now keep wrapper
of SpeechSynthesisUtterance alive until there is no more event to be dispatched. This can happen in two cases:
1. SpeechSynthesisUtterance is not processed yet, and is removed from SpeechSynthesis's queue (i.e. the utterance will
not be processed).
2. SpeechSynthesisUtterance is being processed, and it finishes dispatching end or error event -- these two events are
the last events to be fired on an untterance according to spec: https://wicg.github.io/speech-api/#utterance-events.

* Source/WebCore/Modules/speech/SpeechSynthesis.cpp:
(WebCore::SpeechSynthesis::setPlatformSynthesizer):
(WebCore::SpeechSynthesis::cancel):
(WebCore::SpeechSynthesis::clearUtteranceQueue):
* Source/WebCore/Modules/speech/SpeechSynthesis.h:
* Source/WebCore/Modules/speech/SpeechSynthesisUtterance.cpp:
(WebCore::SpeechSynthesisUtterance::SpeechSynthesisUtterance):
(WebCore::SpeechSynthesisUtterance::eventOccurred):
(WebCore::SpeechSynthesisUtterance::errorEventOccurred):
(WebCore::SpeechSynthesisUtterance::dispatchEventAndUpdateState):
(WebCore::SpeechSynthesisUtterance::setIsActiveForEventDispatch):
(WebCore::SpeechSynthesisUtterance::activeDOMObjectName const):
(WebCore::SpeechSynthesisUtterance::virtualHasPendingActivity const):
* Source/WebCore/Modules/speech/SpeechSynthesisUtterance.h:
* Source/WebCore/Modules/speech/SpeechSynthesisUtterance.idl:

Canonical link: https://commits.webkit.org/266959@main
